### PR TITLE
Added more `NonEmpty` and `Subset` instances, and improve the CTC example.

### PR DESCRIPTION
**Pull request by duvenaud**
The CTC demo was a rat's nest of unsafe indexing arithmetic and would fail for empty inputs.  This PR revisits the CTC demo in light of some of the upgrades to the language, and my under better understanding of functional programming.

Specifically, this PR uses `NonEmpty` instances to replace every instance of `(0@_)` indexing, and manual indexing math, with `first_ix` and `last_ix`.  Along the way I also added a bunch more `NonEmpty` instances to the prelude.

I also took advantage of the new `LogSpace` typeclass to make the math look more like the equations in the paper, hiding away all the `logaddexp` logic and replacing it with simple `+` and '*'.

I also tried to include a `Subset` constraint in the type signature for `ctc`, but ran into the problem that I'd have to create all sorts of instances like `Subset (Fin 10) (Fin 20)`.

I also (with @axch 's help) created two new custom `Ix` instances: `FenceAndPosts` and `AllButFirst`.  To me these are really pleasing ways of factoring out indexing logic into meaningful, reusable, and easily testable chunks.  And together I think they make a compelling demo of how fancy types can help us safely and clearly handle fencepost and off-by-one issues.

The big question to me is if there is some way to automatically generate `NonEmpty` and `Subset` instances for `Fin N` as needed.  One argument for why it's worth adding to the compiler is that we might already want to add special automatic `NonEmpty` instances whenever a `for` loop is entered.

### CODE REVIEW

File: examples/ctc.dx [hunk 1]
```diff
@@ -30,115 +33,88 @@ def prepend {m v} (first: v) (seq: m=>v) : ((Unit|m)=>v) =
...
-    select (o >= 2) (ilabels.s == ilabels.(clipidx _ (o - 2))) False
-
-  safe_idx = \prev s.
-    select (s >= 0) prev.(clipidx _ s) (log 0.0)
-
-  -- Todo: As suggested by Adam Paske, we could get rid of these
-  -- logaddexp calls with a newtype that overloads + and *
-  update = \t prev.
-    case ordinal t == 0 of
+def ctc {vocab time position} [NonEmpty vocab, NonEmpty position, NonEmpty time, Eq time, Eq vocab]
```
Comment (apaszke) [inline examples/ctc.dx:49]: Those NonEmpty constraints are nice, but they might end up being quite annoying if anybody tries to use this code. It might be a better idea to write the "ctc core" implementation that assumes those instances, and then a wrapper around it that does the non-empty check dynamically (basically, cast the arrays to `Unit|Fin something`), and calls the core version once the checks succeed.

File: examples/ctc.dx [hunk 2]
```diff
@@ -30,115 +33,88 @@ def prepend {m v} (first: v) (seq: m=>v) : ((Unit|m)=>v) =
...
 
-def randIdxNoZero (n:Type) [Ix n] (k:Key) : n =
-  unif = rand k
-  from_ordinal n $ unsafe_i_to_n $
-    (1 + (f_to_i (floor ( unif * i_to_f ((n_to_i $ size n) - 1)))))
-
-Vocab = Fin 6
-position = Fin 3
-blank = 0@Vocab
+-- Todo: Produce proofs that Fin N is NonEmpty for all N > 0
```
Comment (apaszke) [inline examples/ctc.dx:91]: I honestly don't see how to do this, so it might be better to leave out the TODO 😕

File: lib/prelude.dx [hunk 3]
```diff
@@ -862,6 +910,24 @@ instance {a b c} [Subset a b, Subset b c] Subset a c
...
+  inject' = \x. Right x
+  project' = \x. case x of
+    Left  x -> Nothing
+    Right x -> Just x
+
+instance {a b} [NonEmpty b, Eq b] Subset a (a & b)
+  inject' = \x. (x, first_ix)
+  project' = \(x, y). case y == first_ix of
+    True  -> Just x
+    False -> Nothing
```
Comment (apaszke) [inline lib/prelude.dx:929]: This is an instance that's a bit too specific to include in the prelude IMO. It would be better to just write those exact injections where you need them.

File: lib/prelude.dx [hunk 4]
```diff
@@ -2469,3 +2535,15 @@ instance {a b} [Ix a, Ix b] Ix (a=>b)
   size = intpow (size b) (size a)
   ordinal             = reversed_digits_to_int
   unsafe_from_ordinal = int_to_reversed_digits
+
+instance {a b} [NonEmpty a, NonEmpty b] NonEmpty (a=>b)
```
Comment (apaszke) [inline lib/prelude.dx:2539]: `a=>b` is never non-empty, since you have `[]`

File: lib/prelude.dx [hunk 5]
```diff
@@ -2469,3 +2535,15 @@ instance {a b} [Ix a, Ix b] Ix (a=>b)
...
+  first_ix = unsafe_from_ordinal _ 0
+
+instance {a b} [NonEmpty a, Ix b, Subset a b] NonEmpty b
+  first_ix = unsafe_from_ordinal _ 0
```
Comment (apaszke) [inline lib/prelude.dx:2543]: I'm a bit afraid this instance will cause some "duplicate instance" errors 😕

File: lib/prelude.dx [hunk 6]
```diff
@@ -2469,3 +2535,15 @@ instance {a b} [Ix a, Ix b] Ix (a=>b)
...
+
+instance {a b} [NonEmpty b, Eq a, Eq b] Subset a (b=>a)
```
Comment (apaszke) [inline lib/prelude.dx:2545]: This feels very arbitrary too. I think the `Subset` is mainly meant for sum-type-like things, not for product types.

**Review by apaszke — COMMENTED** [review]
Nice!

File: examples/ctc.dx (hunk 1 above)
Comment (duvenaud) [reply examples/ctc.dx:49]: Good idea, will do.

File: examples/ctc.dx (hunk 2 above)
Comment (duvenaud) [reply examples/ctc.dx:91]: Yeah, the only way I can think of doing this is either in the compiler.  I also didn't end up using the `Subset` constraint because then I'd need instances like `Subset (Fin 10) (Fin 20)`.  Will do.

File: lib/prelude.dx (hunk 3 above)
Comment (duvenaud) [reply lib/prelude.dx:929]: Oh, I guess you're saying that because the specific injection used here might not match what other users want to do?  I agree, will remove.

This is probably too fine-grained, but I kind of want to split `Subset` into an empty typeclass `IsSubset` and another one like `HasInjection` which contains these methods.  That would let us keep the general assertion in the prelude.  Anyways this is all probably premature without an application.

File: lib/prelude.dx (hunk 6 above)
Comment (duvenaud) [reply lib/prelude.dx:2545]: I agree that the injection strategy is a bit arbitrary, I'll remove.  I got carried away with all the wonderful things I can prove with `NonEmpty` instances!

File: lib/prelude.dx (hunk 5 above)
Comment (duvenaud) [reply lib/prelude.dx:2543]: Good point, will remove.

File: examples/ctc.dx [hunk 7]
```diff
@@ -30,115 +33,115 @@ def prepend {m v} (first: v) (seq: m=>v) : ((Unit|m)=>v) =
...
-    case ordinal t == 0 of
-      True -> log_prob_seq_t0
-      False -> for s.
-        cond = ilabels.s == blank || same_as_last ilabels s
-        labar = logaddexp prev.s (safe_idx prev (n_to_i (ordinal s) - 1))
-        other = logaddexp labar  (safe_idx prev (n_to_i (ordinal s) - 2))
-        ans = select cond labar other
-        ans + normalized_logits.t.(ilabels.s)
+def normalize {n v} [Fractional v, Add v] (xs:n=>v) : n=>v =
+  for i. divide xs.i (sum xs)
```
Comment (axch) [inline examples/ctc.dx:41]: Dex is not yet smart enough yet to lift this `sum xs` out of the enclosing `for` on its own, so as written this does quadratic work.  Recommend computing the sum outside the `for`.

File: examples/ctc.dx [hunk 8]
```diff
@@ -30,115 +33,115 @@ def prepend {m v} (first: v) (seq: m=>v) : ((Unit|m)=>v) =
...
+    (logits: time=>vocab=>LogSpace Float)
+    (labels: position=>vocab)
+      : LogSpace Float =
+  ilabels = prepend_and_interleave blank labels
+  normalized_logits = for t. normalize logits.t
+
+  -- Initialize dynamic programming table.
+  logprob_start_with_blank = normalized_logits.first_ix.blank
+  logprob_of_first_label   = normalized_logits.first_ix.(labels.first_ix)
+  log_prob_seq_t0 = yield_state (for pos. Exp (-10000.0)) \lpRef.
```
Comment (axch) [inline examples/ctc.dx:60]: You can't write -infinity instead of -10000 here?  Is there a NaN bug somewhere in the prelude already?

File: examples/ctc.dx [hunk 9]
```diff
@@ -30,115 +33,115 @@ def prepend {m v} (first: v) (seq: m=>v) : ((Unit|m)=>v) =
...
-
-def ctc {vocab time position} [Eq vocab, Eq position, Eq time]
-      (blank:  vocab)
-      (logits: time=>vocab=>Float)
-      (labels: position=>vocab)
-      : Float =
-  -- Computes log p(labels | logits), marginalizing over possible alignments.
-  -- Todo: remove unnecessary implicit type annotations once
-  -- Dex starts putting implicit types in scope.
+def prepend_and_interleave {m v} (blank:v) (seq: m=>v) : ((Unit|(m & IsBlank))=>v) =
```
Comment (axch) [inline examples/ctc.dx:36]: Not to mess up your architecture here, but do you think this composition might be better served with the `Post` (as in "fence-post") index set from the Prelude?  For an index set `ix`, `Post ix` is the index set of fence-posts around `ix`, supporting functions `left_post : ix -> Post ix` and `right_post : ix -> Post ix`.  Maybe it would make sense to build another index set `FenceAndPosts ix` that iterates over the fences and the posts in the proper order?

**Review by axch — COMMENTED** [review]
The prelude changes LGTM.  Can't speak to the CTC example because I don't really understand it, but left a couple suggestions, two local, one optional.

File: examples/ctc.dx [hunk 10]
```diff
@@ -30,115 +33,115 @@ def prepend {m v} (first: v) (seq: m=>v) : ((Unit|m)=>v) =
...
+      time_size_m1     = unsafe_nat_diff (size time) 1
+      position_size_m1 = unsafe_nat_diff (size position) 1
+
+      vocab'    = (Unit | Fin vocab_size_m1)
+      time'     = (Unit | Fin time_size_m1)
+      position' = (Unit | Fin position_size_m1)
+
+      blank'  = unsafe_from_ordinal vocab' (ordinal blank)
+      logits' = for t:time'. for v:vocab'. logits.(unsafe_from_ordinal _ (ordinal t)).(unsafe_from_ordinal _ (ordinal v))
+      labels' = for p:position'.
```
Comment (apaszke) [inline examples/ctc.dx:111]: you could use `view` loops here

File: lib/prelude.dx [hunk 11]
```diff
@@ -2469,3 +2529,6 @@ instance {a b} [Ix a, Ix b] Ix (a=>b)
   size = intpow (size b) (size a)
   ordinal             = reversed_digits_to_int
   unsafe_from_ordinal = int_to_reversed_digits
+
+instance {a b} [Ix a, NonEmpty b] NonEmpty (a=>b)
```
Comment (apaszke) [inline lib/prelude.dx:2533]: I don't think that `b` needs to be non-empty. `=>` is an integer exponential for which `0^0 = 1` (an empty list is there, even for a `Void` element type)

File: examples/ctc.dx (hunk 7 above)
Comment (duvenaud) [reply examples/ctc.dx:41]: Ah ok, will fix.  Although I guess I view my role as example-writer is partly to write the code that I wish would be fast, in order to spur on compiler development :)

File: examples/ctc.dx (hunk 8 above)
Comment (duvenaud) [reply examples/ctc.dx:60]: Yeah that bugs me too.  But in the main loop we have to multiply two probabilities, which I think is where the `-infinity` can turn into a `NaN`.

File: examples/ctc.dx (hunk 9 above)
Comment (duvenaud) [reply examples/ctc.dx:36]: Oooh, great idea, will try it later today.  I think your will be more readable.

File: examples/ctc.dx (hunk 10 above)
Comment (duvenaud) [reply examples/ctc.dx:111]: Good idea, will do.

File: lib/prelude.dx (hunk 11 above)
Comment (duvenaud) [reply lib/prelude.dx:2533]: Ah, I agree that `0^0 = 1`, and that this possibility isn't covered by my current rule.  Maybe I'll add an instance `instance {a b} [Ix a, Ix b] NonEmpty (Fin 0) (Fin 0)`.

But `0^1 = 0`, i.e. we can't construct an element of type `Unit=>(Fin 0)`.  So to avoid that case, we can't have an unconstrained rule like `instance {a b} [Ix a, Ix b] NonEmpty a b`

File: examples/ctc.dx (hunk 8 above)
Comment (axch) [reply examples/ctc.dx:60]: Shouldn't.  This example works:

```
import stats

Exp (-infinity) * Exp (-infinity)
> (Exp -Infinity)
```

If there's something that doesn't we should probably track it down and fix it.

File: lib/prelude.dx (hunk 11 above)
Comment (apaszke) [reply lib/prelude.dx:2533]: Ah, good point! I think the current rule is sufficient

File: examples/ctc.dx (hunk 8 above)
Comment (duvenaud) [reply examples/ctc.dx]: Thanks to @axch 's help I did track this down and made our `logsumexp` functions handle infinities correctly.

File: examples/ctc.dx [hunk 12]
```diff
@@ -13,89 +13,123 @@ However, Dex's autodiff can produce the backward
...
+  size = 2 * size a + 1
+  ordinal = \i. case i of
+    Fence j -> 2 * ordinal j + 1
+    Posts j -> 2 * ordinal j
+  unsafe_from_ordinal = \o.
+    case is_odd o of
+      True  -> Fence $ unsafe_from_ordinal _ (intdiv2 o)
+      False -> Posts $ unsafe_from_ordinal _ (intdiv2 o)
+
+instance {a} [NonEmpty a] NonEmpty (FenceAndPosts a)
```
Comment (axch) [inline examples/ctc.dx:39]: You don't need `NonEmpty a` here -- `FenceAndPosts` is always non-empty, because there is always at least one `Post`.

File: examples/ctc.dx [hunk 13]
```diff
@@ -13,89 +13,123 @@ However, Dex's autodiff can produce the backward
...
+  size = unsafe_nat_diff (size a) 1
+  ordinal = \(MkAllButFirst i). ordinal i
+  unsafe_from_ordinal = \o. MkAllButFirst $ unsafe_from_ordinal _ o
 
-def prepend_and_interleave {m v} (blank:v) (seq: m=>v) : ((Unit|(m & IsBlank))=>v) =
+instance {a} Subset (AllButFirst a) a
+  inject' = \x. unsafe_from_ordinal _ ((ordinal x) + 1)
+  project' = \x. case (ordinal x) < (size a) of
+    True  -> unsafe_from_ordinal _ (unsafe_nat_diff (ordinal x) 1)
+    False -> Nothing
```
Comment (axch) [inline examples/ctc.dx:58]: This doesn't look right.  You're checking that `ordinal x` is not too big, but then you're subtracting 1 from it?  Shouldn't this be `Nothing` if `ordinal x == 0`?

Let's write a little test of this index set and `Subset` instance (which you can either work into the demo or factor as a separate library with its own tests, or just embed here as an ugly test, possibly at the end).  `project . inject` should be `Just`, and `for x:(AllButFirst a). frob.(inject x)` should iterate over the desired set of elements of `frob : a => Int`.

File: examples/ctc.dx [hunk 14]
```diff
@@ -13,89 +13,123 @@ However, Dex's autodiff can produce the backward
...
+  prob_seq_final = fold prob_seq_at_start \t:(AllButFirst time) prev.
+    for s.
+      prob_stay_same       = prev.s
+      prob_from_blank      = safe_idx_sub prev s 1
+      prob_from_last_token = safe_idx_sub prev s 2
+
+      -- Equation 6 from CTC paper.
+      ans = if ilabels.s == blank || same_as_last s
+        then prob_stay_same + prob_from_blank
+        else prob_stay_same + prob_from_blank + prob_from_last_token
```
Comment (axch) [inline examples/ctc.dx:114]: I didn't read the paper, but it looks weird to have `prob_stay_same` appear if the label is not the same?  If that's intended, maybe a brief comment would help?

File: lib/stats.dx [hunk 15]
```diff
@@ -16,11 +32,25 @@ instance {f} [Sub f] Fractional (LogSpace f)
 instance Arbitrary (LogSpace Float)
   arb = \k. Exp (randn k)
 
+def is_finite {f} [Fractional f, Sub f, Mul f, Ord f] (x:f) : Bool =
+  -- Todo: Make polymorphic versions of these and put them in the prelude.
+  infinity = (divide one zero)
+  neg_infinity = sub zero infinity
+  not (x == infinity || x == neg_infinity)
```
Comment (axch) [inline lib/stats.dx:39]: Conventionally, is_finite nan should return False.  NaN can be checked for because it's the only floating-point number x satisfying `not (x == x)`.  Could you add that condition, and throw in a unit test for is_finite to make sure it stays correct in the future?

**Review by axch — CHANGES_REQUESTED** [review]
Nice!  We're almost at the state where if you're happy then I'm happy.  Just a couple small comments about the indexing and the numerics.

File: examples/ctc.dx (hunk 12 above)
Comment (duvenaud) [reply examples/ctc.dx:39]: Oh yeah, good point.

File: examples/ctc.dx (hunk 13 above)
Comment (duvenaud) [reply examples/ctc.dx:58]: Great point, thanks for catching this.  I'll add a test to this example for now, and if we ever have another use for this index set, we can move it to the prelude or a library as you suggest.

File: examples/ctc.dx (hunk 14 above)
Comment (duvenaud) [reply examples/ctc.dx:114]: Good point, these are poorly named.  The paper is just a nest of equations without any names.  `same_as_last` and `stay_same` have a confusing collision because there are two distinct notions of time (one for the tokens and one for the sounds).  I'll take another stab at these names.

File: lib/stats.dx (hunk 15 above)
Comment (duvenaud) [reply lib/stats.dx:39]: Yes, will do.

File: examples/ctc.dx (hunk 12 above)
Comment (duvenaud) [reply examples/ctc.dx]: Done.

File: examples/ctc.dx (hunk 13 above)
Comment (duvenaud) [reply examples/ctc.dx]: Fixed and tests added.

File: examples/ctc.dx (hunk 14 above)
Comment (duvenaud) [reply examples/ctc.dx]: Renamed to hopefully be clearer.

File: lib/stats.dx (hunk 15 above)
Comment (duvenaud) [reply lib/stats.dx]: I was using this function in a way where I didn't want to include `nan`.  So I changed from `in_finite` to `is_infinite`, which doesn't include `nan`.  I added a comment and some tests to make the intended behavior clear.

**Comment by duvenaud** [comment]
I'm happy to merge now.  Thanks for the feedback!
